src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormattingPass.cs

```diff
@@ -0,0 +1,172 @@
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<HtmlFormattingPass>();
+        }
+
+        // Run after the HTML formatter pass.
+        public override int Order => DefaultOrder - 5;
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result)
+        {
+            if (context.IsFormatOnType || result.Kind != RazorLanguageKind.Razor)
+            {
+                // We don't want to handle OnTypeFormatting here.
+                return result;
+            }
+
+            // Apply previous edits if any.
+            var originalText = context.CodeDocument.GetSourceText();
+            var changedText = originalText;
+            var changedContext = context;
+            if (result.Edits.Length > 0)
+            {
+                var changes = result.Edits.Select(e => e.AsTextChange(originalText)).ToArray();
+                changedText = changedText.WithChanges(changes);
+                changedContext = await context.WithTextAsync(changedText);
+            }
+
+            // Apply original C# edits
+            var csharpEdits = await FormatCSharp(changedContext);
+            if (csharpEdits.Count > 0)
+            {
+                var csharpChanges = csharpEdits.Select(c => c.AsTextChange(changedText));
+                changedText = changedText.WithChanges(csharpChanges);
+                changedContext = await changedContext.WithTextAsync(changedText);
+            }
+
+            // Now, for each affected line in the edited version of the document, remove x amount of spaces
+            // at the front to account for extra indentation applied by the C# formatter.
+            // This should be based on context.
+            // For instance, lines inside @code/@functions block should be reduced one level
+            // and lines inside @{} should be reduced by two levels.
+            var indentationChanges = AdjustCSharpIndentation(changedContext, 0, changedText.Lines.Count - 1);
+
+            if (indentationChanges.Count > 0)
+            {
+                // Apply the edits that remove indentation.
+                changedText = changedText.WithChanges(indentationChanges);
+                changedContext = await changedContext.WithTextAsync(changedText);
+            }
+
+            // We make an optimistic attempt at fixing corner cases.
+            changedText = CleanupDocument(changedContext);
+
+            var finalChanges = SourceTextDiffer.GetMinimalTextChanges(originalText, changedText, lineDiffOnly: false);
+            var finalEdits = finalChanges.Select(f => f.AsTextEdit(originalText)).ToArray();
+
+            return new FormattingResult(finalEdits);
+        }
+
+        private async Task<List<TextEdit>> FormatCSharp(FormattingContext context)
+        {
+            var sourceText = context.CodeDocument.GetSourceText();
```

Comment:
Do you feel the context.GetSourceText variant is less clear/valuable? If that's the case we should consider removing it.

Reply:
I forgot about it :D. Fixed it in other places as well